Mathlib/Analysis/Calculus/MeanValue.lean

```diff
@@ -0,0 +1,1365 @@
+    (hderiv : ∀ x ∈ Ico a b, derivWithin f (Icc a b) x = 0) : ∀ x ∈ Icc a b, f x = f a := by
+  have H : ∀ x ∈ Ico a b, ‖derivWithin f (Icc a b) x‖ ≤ 0 := by
+    simpa only [norm_le_zero_iff] using fun x hx => hderiv x hx
+  simpa only [zero_mul, norm_le_zero_iff, sub_eq_zero] using fun x hx =>
+    norm_image_sub_le_of_norm_deriv_le_segment hdiff H x hx
+#align constant_of_deriv_within_zero constant_of_derivWithin_zero
+
+variable {f' g : ℝ → E}
+
+/-- If two continuous functions on `[a, b]` have the same right derivative and are equal at `a`,
+  then they are equal everywhere on `[a, b]`. -/
+theorem eq_of_has_deriv_right_eq (derivf : ∀ x ∈ Ico a b, HasDerivWithinAt f (f' x) (Ici x) x)
+    (derivg : ∀ x ∈ Ico a b, HasDerivWithinAt g (f' x) (Ici x) x) (fcont : ContinuousOn f (Icc a b))
+    (gcont : ContinuousOn g (Icc a b)) (hi : f a = g a) : ∀ y ∈ Icc a b, f y = g y := by
+  simp only [← @sub_eq_zero _ _ (f _)] at hi⊢
+  exact hi ▸ constant_of_has_deriv_right_zero (fcont.sub gcont) fun y hy => by
+    simpa only [sub_self] using (derivf y hy).sub (derivg y hy)
+#align eq_of_has_deriv_right_eq eq_of_has_deriv_right_eq
+
+/-- If two differentiable functions on `[a, b]` have the same derivative within `[a, b]` everywhere
+  on `[a, b)` and are equal at `a`, then they are equal everywhere on `[a, b]`. -/
+theorem eq_of_derivWithin_eq (fdiff : DifferentiableOn ℝ f (Icc a b))
+    (gdiff : DifferentiableOn ℝ g (Icc a b))
+    (hderiv : EqOn (derivWithin f (Icc a b)) (derivWithin g (Icc a b)) (Ico a b)) (hi : f a = g a) :
+    ∀ y ∈ Icc a b, f y = g y := by
+  have A : ∀ y ∈ Ico a b, HasDerivWithinAt f (derivWithin f (Icc a b) y) (Ici y) y := fun y hy =>
+    (fdiff y (mem_Icc_of_Ico hy)).hasDerivWithinAt.nhdsWithin (Icc_mem_nhdsWithin_Ici hy)
+  have B : ∀ y ∈ Ico a b, HasDerivWithinAt g (derivWithin g (Icc a b) y) (Ici y) y := fun y hy =>
+    (gdiff y (mem_Icc_of_Ico hy)).hasDerivWithinAt.nhdsWithin (Icc_mem_nhdsWithin_Ici hy)
+  exact
+    eq_of_has_deriv_right_eq A (fun y hy => (hderiv hy).symm ▸ B y hy) fdiff.continuousOn
+      gdiff.continuousOn hi
+#align eq_of_deriv_within_eq eq_of_derivWithin_eq
+
+end
+
+/-!
+### Vector-valued functions `f : E → G`
+
+Theorems in this section work both for real and complex differentiable functions. We use assumptions
+`[IsROrC 𝕜] [NormedSpace 𝕜 E] [NormedSpace 𝕜 G]` to achieve this result. For the domain `E` we
+also assume `[NormedSpace ℝ E]` to have a notion of a `Convex` set. -/
+
+section
+
+variable {𝕜 G : Type _} [IsROrC 𝕜] [NormedSpace 𝕜 E] [NormedAddCommGroup G] [NormedSpace 𝕜 G]
+
+namespace Convex
+
+variable {f g : E → G} {C : ℝ} {s : Set E} {x y : E} {f' g' : E → E →L[𝕜] G} {φ : E →L[𝕜] G}
+
+open AffineMap in
+/-- The mean value theorem on a convex set: if the derivative of a function is bounded by `C`, then
+the function is `C`-Lipschitz. Version with `has_fderiv_within`. -/
+theorem norm_image_sub_le_of_norm_has_fderiv_within_le
```

Comment:
```suggestion
theorem norm_image_sub_le_of_norm_hasFderivWithin_le
```
Various lemma names still have to be fixed, I was a bit too hasty with merging

Reply:
Should be done now.